diff --git a/Applications/TransferLearning/Draw_Classification/jni/main.cpp b/Applications/TransferLearning/Draw_Classification/jni/main.cpp
index fa0c749..86a2df8 100644
--- a/Applications/TransferLearning/Draw_Classification/jni/main.cpp
+++ b/Applications/TransferLearning/Draw_Classification/jni/main.cpp
@@ -405,6 +405,7 @@ int trainModel(const char *config) {
 #if defined(__TIZEN__)
 void sink_cb(const ml_tensors_data_h data, const ml_tensors_info_h info,
              void *user_data) {
+  static int PREDICTION_THRESHOLD = 0.7;
   static int test_file_idx = 0;
   int status = ML_ERROR_NONE;
   ml_tensor_dimension dim;
@@ -421,14 +422,19 @@ void sink_cb(const ml_tensors_data_h data, const ml_tensors_info_h info,
     return;
 
   for (int i = 0; i < LABEL_SIZE; i++) {
-    if (raw_data[i] > max_val) {
+    if (raw_data[i] > max_val && raw_data[i] > PREDICTION_THRESHOLD) {
       max_val = raw_data[i];
       max_idx = i;
     }
   }
 
-  std::cout << "Label for test file test" << test_file_idx
-            << ".bmp = " << label_names[max_idx] << std::endl;
+  std::cout << "Label for test file test" << test_file_idx + 1
+            << ".bmp = ";
+  if (max_idx >= 0) {
+    std::cout << label_names[max_idx]  << " with softmax value " << max_val << std::endl;
+  } else {
+    std::cout << "cannot be predicted with enough confidence.";
+  }
   test_file_idx += 1;
 }
 #endif
